smolyay/surrogate.py

```diff
@@ -0,0 +1,309 @@
+
+    def train(self, function, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to the function.
+
+        Parameters
+        ----------
+        function: callabe
+            Function to be approximated.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition.
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+        """
+        data = [function(point) for point in self.points]
+        self.train_from_data(data, linear_solver)
+
+    def train_from_data(self, data, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to data.
+
+        Parameters
+        ----------
+        data: list
+            function at grid points.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Raises
+        ------
+        IndexError
+            Data should be of length of the grid points.
+        ValueError
+            Solver should be selected between defined methods.
+        """
+        data = numpy.array(data, copy=False, ndmin=1)
+        if data.shape != (self.points.shape[0],):
+            raise IndexError("Data must be same length as grid points.")
+        basis_matrix = self._make_basis_matrix()
+        if linear_solver == 'lu':
+            self._coefficients = numpy.linalg.solve(
+                basis_matrix,
+                data)
+        elif linear_solver == 'inv':
+            self._coefficients = numpy.dot(numpy.linalg.inv(
+                basis_matrix),
+                data)
+        elif linear_solver == 'lstsq':
+            self._coefficients = numpy.linalg.lstsq(basis_matrix,
+                                                    data,
+                                                    rcond=None)[0]
+        else:
+            raise ValueError('Solver not recognized')
+
+    def _make_basis_matrix(self):
+        """Generate basis matrix.
+
+        Parameters
+        ----------
+        points: list
+            Grid points.
+        basis_functions: list
+            Basis function of grid points.
+
+        Returns
+        -------
+        numpy.ndarray
+            Basis matrix.
+        """
+        points, basis_functions = self.grid.points, self.grid.basis_functions
+        basis_matrix = numpy.zeros((len(points), len(points)))
+        for i, point in enumerate(points):
+            for j, basis in enumerate(basis_functions):
+                if self.dimension > 1:
+                    value = numpy.prod([f(x) for x, f in zip(point, basis)])
+                else:
+                    value = basis(point)
+                basis_matrix[i, j] = value
+        return basis_matrix
+
+    def _mapdomain(self, x, old, new):
```

Comment:
The only part where this actually uses `self` is to check the dimensionality, so let's remove that and make this a static method:
```suggestion
    @staticmethod
    def _mapdomain(x, old, new):
```